pkg/plugins/mco/mco_plugin.go

```diff
@@ -0,0 +1,203 @@
+func init() {
+	Plugin = McoPlugin{
+		PluginName:  "mco_plugin",
+		SpecVersion: "1.0",
+	}
+
+	var config *rest.Config
+	var err error
+	kubeconfig := os.Getenv("KUBECONFIG")
+	if kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		// creates the in-cluster config
+		config, err = rest.InClusterConfig()
+	}
+
+	if err != nil {
+		panic(err.Error())
+	}
+	kubeclient = kubernetes.NewForConfigOrDie(config)
+}
+
+// Name returns the name of the plugin
+func (p *McoPlugin) Name() string {
+	return p.PluginName
+}
+
+// Spec returns the version of the spec expected by the plugin
+func (p *McoPlugin) Spec() string {
+	return p.SpecVersion
+}
+
+// OnNodeStateAdd Invoked when SriovNetworkNodeState CR is created, return if need dain and/or reboot node
+func (p *McoPlugin) OnNodeStateAdd(state *sriovnetworkv1.SriovNetworkNodeState) (needDrain bool, needReboot bool, err error) {
+	glog.Info("mco-plugin OnNodeStateAdd()")
+	nodeName = state.GetName()
+	return p.OnNodeStateChange(nil, state)
+}
+
+// OnNodeStateChange Invoked when SriovNetworkNodeState CR is updated, return if need dain and/or reboot node
+func (p *McoPlugin) OnNodeStateChange(old, new *sriovnetworkv1.SriovNetworkNodeState) (needDrain bool, needReboot bool, err error) {
+	glog.Info("mco-plugin OnNodeStateChange()")
+	switchdevConfigured = false
+	for _, iface := range new.Spec.Interfaces {
+		if iface.EswitchMode == sriovnetworkv1.ESWITCHMODE_SWITCHDEV {
+			switchdevConfigured = true
+		}
+	}
+
+	var update, remove bool
+	if update, remove, err = writeSwitchdevConfFile(new); err != nil {
+		glog.Errorf("mco-plugin OnNodeStateChange():fail to update switchdev.conf file: %v", err)
+		return
+	}
+	if remove {
+		needDrain = true
+		return
+	}
+	if update {
+		if _, err = os.Stat(switchdevUnitPath); err != nil {
+			if os.IsNotExist(err) {
+				glog.Info("mco-plugin OnNodeStateChange(): the latest MachineConfig has not been applied")
+				needDrain = true
+				err = nil
+				return
+			}
+			return
+		}
+		// node is already in the offload MCP
+		glog.Info("mco-plugin OnNodeStateChange(): need reboot node to use the up-to-date switchdev.conf")
+		needDrain = true
+		needReboot = true
+		return
+	}
+	return
+}
+
+// Apply config change
+func (p *McoPlugin) Apply() error {
+	glog.Info("mco-plugin Apply()")
+	node, err := kubeclient.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	label := nodeLabelPrefix + controllers.HwOffloadNodeLabel
+	if switchdevConfigured {
+		if _, ok := node.Labels[label]; !ok {
+			glog.Infof("Move node %s into HW offload MachineConfigPool", node.Name)
+			mergePatch, _ := json.Marshal(map[string]interface{}{
+				"metadata": map[string]interface{}{
+					"labels": map[string]interface{}{
+						label: "",
+					},
+				},
+			})
+			kubeclient.CoreV1().Nodes().Patch(context.Background(), nodeName, types.MergePatchType, mergePatch, metav1.PatchOptions{})
+		}
```

Comment:
return or add `else` here? otherwise the next glog message will always be printed.

Reply:
fixed.